go/border/rpkt/path.go

```diff
@@ -250,58 +253,77 @@ func (rp *RtrPkt) hopFVerFromRaw(offset int) common.RawBytes {
 	return ans
 }
 
-// IncPath increments the packet's path, if any.
-func (rp *RtrPkt) IncPath() *common.Error {
+// IncPath increments the packet's path, if any. The bool return value is set
+// to true if the segment changes (specifically if the packet's metadata is
+// updated to the new segment).
+func (rp *RtrPkt) IncPath() (bool, *common.Error) {
 	if rp.infoF == nil {
 		// Path is empty, nothing to increment.
-		return nil
+		return false, nil
 	}
 	var err *common.Error
 	var hopF *spath.HopField
 	// Initialize to the current InfoF and offset values.
 	infoF := rp.infoF
 	iOff := rp.CmnHdr.CurrInfoF
 	hOff := rp.CmnHdr.CurrHopF
+	vOnly := 0
+	origUp := *rp.upFlag
 	for {
 		hOff += spath.HopFieldLength
 		if hOff-iOff > infoF.Hops*spath.HopFieldLength {
 			// Passed end of current segment, switch to next segment, and read
 			// the new Info Field.
 			iOff = hOff
 			if infoF, err = spath.InfoFFromRaw(rp.Raw[iOff:]); err != nil {
-				return err
+				// Still return false as the metadata hasn't been updated to the new segment.
+				return false, err
 			}
 			continue
 		}
 		// Read new Hop Field
 		if hopF, err = spath.HopFFromRaw(rp.Raw[hOff:]); err != nil {
-			return err
+			return false, err
 		}
 		// Find first non-verify-only Hop Field.
 		if !hopF.VerifyOnly {
 			break
 		}
+		vOnly++
 	}
 	if hOff > rp.CmnHdr.HdrLen {
-		return common.NewError("New HopF offset > header length", "max", rp.CmnHdr.HdrLen,
+		return false, common.NewError("New HopF offset > header length", "max", rp.CmnHdr.HdrLen,
 			"actual", hOff)
 	}
 	// Update common header, and packet's InfoF/HopF fields.
+	segChgd := iOff != rp.CmnHdr.CurrInfoF
 	rp.CmnHdr.UpdatePathOffsets(rp.Raw, iOff, hOff)
 	rp.infoF = infoF
 	rp.hopF = hopF
-	rp.upFlag = nil
 	rp.IncrementedPath = true
-	// Extract new Up flag, in case the Info Field has changed.
-	if _, err = rp.UpFlag(); err != nil {
-		return err
+	if segChgd {
+		// Extract new Up flag.
+		rp.upFlag = nil
+		if _, err = rp.UpFlag(); err != nil {
+			return segChgd, err
+		}
 	}
 	// Extract the next interface ID.
 	rp.ifNext = nil
 	if _, err = rp.IFNext(); err != nil {
-		return err
+		return segChgd, err
+	}
+	// Check that there's no VERIFY_ONLY fields in the middle of a segment.
+	if vOnly > 0 && !segChgd {
+		sdata := scmp.NewErrData(scmp.C_Path, scmp.T_P_BadHopField, rp.mkInfoPathOffsets())
+		return segChgd, common.NewError("VERIFY_ONLY in middle of segment", sdata)
+	}
+	// Check that the segment didn't change from a down-segment to an up-segment.
```

Comment:
Good. We also need to check that if we are not a core node, then we do not switch from up to up or from down to down.